traits/trait_list_object.py

```diff
@@ -53,465 +90,765 @@ def __init__(self, index=0, removed=None, added=None):
+            index : slice
+                The range of the slice will be 0 to len(self).
+            added : list
+                The empty list [].
+            removed : list
+                The removed items.
 
-                list.extend(self, xlist)
+        """
+        index = slice(0, len(self))
+        removed = self.copy()
+        added = []
+        self.validate(index, removed, added)
+        super().clear()
+
+        if added != removed:
+            self.notify(index, removed, [])
+
+    def pop(self, index=-1):
+        """ Remove and return item at index (default last).
+
+        Parameters
+        ----------
+        index: int
+            Index at which item has to be removed.
+
+        Returns
+        -------
+        item : An item in the list at given index, last item if no index given.
+
+        Raises
+        ------
+        IndexError
+            If the index is out of range.
+
+        Notes
+        -----
+        Parameters in the notification:
+            index : int
+                The normalized index between 0 and len(self).
+            added : list
+                Will be []
+            removed : list
+                The removed item, as a list.
+
+        """
 
-                if (self.name_items is not None) and (len(xlist) != 0):
-                    self._send_trait_items_event(
-                        self.name_items,
-                        TraitListEvent(len(self) - len(xlist), None, xlist),
-                        trait.items_event(),
-                    )
+        added = []
+        removed = self._get_removed(index, default=[])
+        norm_index = self._normalize_index(index)
+        self.validate(norm_index, removed, added)
 
-                return
+        removed_item = super().pop(index)
 
-            except TraitError as excp:
-                excp.set_prefix("The elements of the")
-                raise excp
+        if added != removed:
+            self.notify(norm_index, removed, added)
 
-        self.len_error(len(self) + len(xlist))
+        return removed_item
 
     def remove(self, value):
-        trait = getattr(self, "trait", None)
-        if trait is None:
-            list.remove(self, value)
-            return
-        if trait.minlen < len(self):
-            try:
-                index = self.index(value)
-                removed = [self[index]]
-            except Exception:
-                pass
-
-            list.remove(self, value)
-
-            if self.name_items is not None:
-                self._send_trait_items_event(
-                    self.name_items, TraitListEvent(index, removed)
-                )
-        elif len(self) == 0:
-            # Let whatever system error (ValueError) should be raised be
-            # raised.
-            list.remove(self, value)
-        else:
-            self.len_error(len(self) - 1)
+        """ Remove first occurrence of value.
+
+        Parameters
+        ----------
+        value : Any
```

Comment:
"any" rather than "Any" here. As discussed before, we don't want to use either Trait types or typing types in this context: trait types aren't a good match, and `typing` types risk confusion.